<a href="https://colab.research.google.com/github/anshikaj1/CodSoft_MACHINE_LEARNING/blob/main/credit_card_fraud_detection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#importing libraries and modules for logistic regression
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split as tts
from sklearn.linear_model import LogisticRegression
from sklearn. metrics import accuracy_score as acs


In [ ]:
#loading dataset
#pandas dataframes
fraud_test_data= pd.read_csv('/content/fraudTest.csv')
a= fraud_test_data

In [ ]:
#first five rows of the dataset will be displayed
a.head()

Unnamed: 0 trans_date_trans_time            cc_num  \
0           0   2020-06-21 12:14:25  2291163933867244   
1           1   2020-06-21 12:14:33  3573030041201292   
2           2   2020-06-21 12:14:53  3598215285024754   
3           3   2020-06-21 12:15:15  3591919803438423   
4           4   2020-06-21 12:15:17  3526826139003047   

                               merchant        category    amt   first  \
0                 fraud_Kirlin and Sons   personal_care   2.86    Jeff   
1                  fraud_Sporer-Keebler   personal_care  29.84  Joanne   
2  fraud_Swaniawski, Nitzsche and Welch  health_fitness  41.28  Ashley   
3                     fraud_Haley Group        misc_pos  60.05   Brian   
4                 fraud_Johnston-Casper          travel   3.19  Nathan   

       last gender                       street  ...      lat      long  \
0   Elliott      M            351 Darlene Green  ...  33.9659  -80.9355   
1  Williams      F             3638 Marsh Union  ...  40.3207 -110.4360   
2     Lopez      F         9333 Valentine Point  ...  40.6729  -73.5365   
3  Williams      M  32941 Krystal Mill Apt. 552  ...  28.5697  -80.8191   
4    Massey      M     5783 Evan Roads Apt. 465  ...  44.2529  -85.0170   

   city_pop                     job         dob  \
0    333497     Mechanical engineer  1968-03-19   
1       302  Sales professional, IT  1990-01-17   
2     34496       Librarian, public  1970-10-21   
3     54767            Set designer  1987-07-25   
4      1126      Furniture designer  1955-07-06   

                          trans_num   unix_time  merch_lat  merch_long  \
0  2da90c7d74bd46a0caf3777415b3ebd3  1371816865  33.986391  -81.200714   
1  324cc204407e99f51b0d6ca0055005e7  1371816873  39.450498 -109.960431   
2  c81755dbbbea9d5c77f094348a7579be  1371816893  40.495810  -74.196111   
3  2159175b9efe66dc301f149d3d5abf8c  1371816915  28.812398  -80.883061   
4  57ff021bd3f328f8738bb535c302a31b  1371816917  44.959148  -85.884734   

   is_fraud  
0         0  
1         0  
2         0  
3         0  
4         0  

[5 rows x 23 columns]

In [ ]:
#last five rows of the dataframes will be displayed
a.tail()

Unnamed: 0 trans_date_trans_time            cc_num  \
555714      555714   2020-12-31 23:59:07    30560609640617   
555715      555715   2020-12-31 23:59:09  3556613125071656   
555716      555716   2020-12-31 23:59:15  6011724471098086   
555717      555717   2020-12-31 23:59:24     4079773899158   
555718      555718   2020-12-31 23:59:34  4170689372027579   

                     merchant        category     amt    first     last  \
555714  fraud_Reilly and Sons  health_fitness   43.77  Michael    Olson   
555715   fraud_Hoppe-Parisian       kids_pets  111.84     Jose  Vasquez   
555716        fraud_Rau-Robel       kids_pets   86.88      Ann   Lawson   
555717  fraud_Breitenberg LLC          travel    7.99     Eric  Preston   
555718      fraud_Dare-Marvin   entertainment   38.13   Samuel     Frey   

       gender                      street  ...      lat      long  city_pop  \
555714      M         558 Michael Estates  ...  40.4931  -91.8912       519   
555715      M         572 Davis Mountains  ...  29.0393  -95.4401     28739   
555716      F  144 Evans Islands Apt. 683  ...  46.1966 -118.9017      3684   
555717      M  7020 Doyle Stream Apt. 951  ...  44.6255 -116.4493       129   
555718      M    830 Myers Plaza Apt. 384  ...  35.6665  -97.4798    116001   

                   job         dob                         trans_num  \
555714    Town planner  1966-02-13  9b1f753c79894c9f4b71f04581835ada   
555715  Futures trader  1999-12-27  2090647dac2c89a1d86c514c427f5b91   
555716        Musician  1981-11-29  6c5b7c8add471975aa0fec023b2e8408   
555717    Cartographer  1965-12-15  14392d723bb7737606b2700ac791b7aa   
555718     Media buyer  1993-05-10  1765bb45b3aa3224b4cdcb6e7a96cee3   

         unix_time  merch_lat  merch_long  is_fraud  
555714  1388534347  39.946837  -91.333331         0  
555715  1388534349  29.661049  -96.186633         0  
555716  1388534355  46.658340 -119.715054         0  
555717  1388534364  44.470525 -117.080888         0  
555718  1388534374  36.210097  -97.036372         0  

[5 rows x 23 columns]

In [ ]:
#structure and information about the dataset
a.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 555719 entries, 0 to 555718
Data columns (total 23 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   Unnamed: 0             555719 non-null  int64  
 1   trans_date_trans_time  555719 non-null  object 
 2   cc_num                 555719 non-null  int64  
 3   merchant               555719 non-null  object 
 4   category               555719 non-null  object 
 5   amt                    555719 non-null  float64
 6   first                  555719 non-null  object 
 7   last                   555719 non-null  object 
 8   gender                 555719 non-null  object 
 9   street                 555719 non-null  object 
 10  city                   555719 non-null  object 
 11  state                  555719 non-null  object 
 12  zip                    555719 non-null  int64  
 13  lat                    555719 non-null  float64
 14  long                   555719 non-nu

In [ ]:
#no. of missing values(NULL VALUES) in the dataset in each column
a.isnull().sum()

Unnamed: 0               0
trans_date_trans_time    0
cc_num                   0
merchant                 0
category                 0
amt                      0
first                    0
last                     0
gender                   0
street                   0
city                     0
state                    0
zip                      0
lat                      0
long                     0
city_pop                 0
job                      0
dob                      0
trans_num                0
unix_time                0
merch_lat                0
merch_long               0
is_fraud                 0
dtype: int64

In [ ]:
#the values which are NOT NULL
a[a.notnull()]

Unnamed: 0 trans_date_trans_time            cc_num  \
0                0   2020-06-21 12:14:25  2291163933867244   
1                1   2020-06-21 12:14:33  3573030041201292   
2                2   2020-06-21 12:14:53  3598215285024754   
3                3   2020-06-21 12:15:15  3591919803438423   
4                4   2020-06-21 12:15:17  3526826139003047   
...            ...                   ...               ...   
555714      555714   2020-12-31 23:59:07    30560609640617   
555715      555715   2020-12-31 23:59:09  3556613125071656   
555716      555716   2020-12-31 23:59:15  6011724471098086   
555717      555717   2020-12-31 23:59:24     4079773899158   
555718      555718   2020-12-31 23:59:34  4170689372027579   

                                    merchant        category     amt    first  \
0                      fraud_Kirlin and Sons   personal_care    2.86     Jeff   
1                       fraud_Sporer-Keebler   personal_care   29.84   Joanne   
2       fraud_Swaniawski, Nitzsche and Welch  health_fitness   41.28   Ashley   
3                          fraud_Haley Group        misc_pos   60.05    Brian   
4                      fraud_Johnston-Casper          travel    3.19   Nathan   
...                                      ...             ...     ...      ...   
555714                 fraud_Reilly and Sons  health_fitness   43.77  Michael   
555715                  fraud_Hoppe-Parisian       kids_pets  111.84     Jose   
555716                       fraud_Rau-Robel       kids_pets   86.88      Ann   
555717                 fraud_Breitenberg LLC          travel    7.99     Eric   
555718                     fraud_Dare-Marvin   entertainment   38.13   Samuel   

            last gender                       street  ...      lat      long  \
0        Elliott      M            351 Darlene Green  ...  33.9659  -80.9355   
1       Williams      F             3638 Marsh Union  ...  40.3207 -110.4360   
2          Lopez      F         9333 Valentine Point  ...  40.6729  -73.5365   
3       Williams      M  32941 Krystal Mill Apt. 552  ...  28.5697  -80.8191   
4         Massey      M     5783 Evan Roads Apt. 465  ...  44.2529  -85.0170   
...          ...    ...                          ...  ...      ...       ...   
555714     Olson      M          558 Michael Estates  ...  40.4931  -91.8912   
555715   Vasquez      M          572 Davis Mountains  ...  29.0393  -95.4401   
555716    Lawson      F   144 Evans Islands Apt. 683  ...  46.1966 -118.9017   
555717   Preston      M   7020 Doyle Stream Apt. 951  ...  44.6255 -116.4493   
555718      Frey      M     830 Myers Plaza Apt. 384  ...  35.6665  -97.4798   

        city_pop                     job         dob  \
0         333497     Mechanical engineer  1968-03-19   
1            302  Sales professional, IT  1990-01-17   
2          34496       Librarian, public  1970-10-21   
3          54767            Set designer  1987-07-25   
4           1126      Furniture designer  1955-07-06   
...          ...                     ...         ...   
555714       519            Town planner  1966-02-13   
555715     28739          Futures trader  1999-12-27   
555716      3684                Musician  1981-11-29   
555717       129            Cartographer  1965-12-15   
555718    116001             Media buyer  1993-05-10   

                               trans_num   unix_time  merch_lat  merch_long  \
0       2da90c7d74bd46a0caf3777415b3ebd3  1371816865  33.986391  -81.200714   
1       324cc204407e99f51b0d6ca0055005e7  1371816873  39.450498 -109.960431   
2       c81755dbbbea9d5c77f094348a7579be  1371816893  40.495810  -74.196111   
3       2159175b9efe66dc301f149d3d5abf8c  1371816915  28.812398  -80.883061   
4       57ff021bd3f328f8738bb535c302a31b  1371816917  44.959148  -85.884734   
...                                  ...         ...        ...         ...   
555714  9b1f753c79894c9f4b71f04581835ada  1388534347  39.946837  -91.333331   
555715  2090647dac2c89a1

In [ ]:
#dropping all NULL VALUES
a.dropna()

Unnamed: 0 trans_date_trans_time            cc_num  \
0                0   2020-06-21 12:14:25  2291163933867244   
1                1   2020-06-21 12:14:33  3573030041201292   
2                2   2020-06-21 12:14:53  3598215285024754   
3                3   2020-06-21 12:15:15  3591919803438423   
4                4   2020-06-21 12:15:17  3526826139003047   
...            ...                   ...               ...   
555714      555714   2020-12-31 23:59:07    30560609640617   
555715      555715   2020-12-31 23:59:09  3556613125071656   
555716      555716   2020-12-31 23:59:15  6011724471098086   
555717      555717   2020-12-31 23:59:24     4079773899158   
555718      555718   2020-12-31 23:59:34  4170689372027579   

                                    merchant        category     amt    first  \
0                      fraud_Kirlin and Sons   personal_care    2.86     Jeff   
1                       fraud_Sporer-Keebler   personal_care   29.84   Joanne   
2       fraud_Swaniawski, Nitzsche and Welch  health_fitness   41.28   Ashley   
3                          fraud_Haley Group        misc_pos   60.05    Brian   
4                      fraud_Johnston-Casper          travel    3.19   Nathan   
...                                      ...             ...     ...      ...   
555714                 fraud_Reilly and Sons  health_fitness   43.77  Michael   
555715                  fraud_Hoppe-Parisian       kids_pets  111.84     Jose   
555716                       fraud_Rau-Robel       kids_pets   86.88      Ann   
555717                 fraud_Breitenberg LLC          travel    7.99     Eric   
555718                     fraud_Dare-Marvin   entertainment   38.13   Samuel   

            last gender                       street  ...      lat      long  \
0        Elliott      M            351 Darlene Green  ...  33.9659  -80.9355   
1       Williams      F             3638 Marsh Union  ...  40.3207 -110.4360   
2          Lopez      F         9333 Valentine Point  ...  40.6729  -73.5365   
3       Williams      M  32941 Krystal Mill Apt. 552  ...  28.5697  -80.8191   
4         Massey      M     5783 Evan Roads Apt. 465  ...  44.2529  -85.0170   
...          ...    ...                          ...  ...      ...       ...   
555714     Olson      M          558 Michael Estates  ...  40.4931  -91.8912   
555715   Vasquez      M          572 Davis Mountains  ...  29.0393  -95.4401   
555716    Lawson      F   144 Evans Islands Apt. 683  ...  46.1966 -118.9017   
555717   Preston      M   7020 Doyle Stream Apt. 951  ...  44.6255 -116.4493   
555718      Frey      M     830 Myers Plaza Apt. 384  ...  35.6665  -97.4798   

        city_pop                     job         dob  \
0         333497     Mechanical engineer  1968-03-19   
1            302  Sales professional, IT  1990-01-17   
2          34496       Librarian, public  1970-10-21   
3          54767            Set designer  1987-07-25   
4           1126      Furniture designer  1955-07-06   
...          ...                     ...         ...   
555714       519            Town planner  1966-02-13   
555715     28739          Futures trader  1999-12-27   
555716      3684                Musician  1981-11-29   
555717       129            Cartographer  1965-12-15   
555718    116001             Media buyer  1993-05-10   

                               trans_num   unix_time  merch_lat  merch_long  \
0       2da90c7d74bd46a0caf3777415b3ebd3  1371816865  33.986391  -81.200714   
1       324cc204407e99f51b0d6ca0055005e7  1371816873  39.450498 -109.960431   
2       c81755dbbbea9d5c77f094348a7579be  1371816893  40.495810  -74.196111   
3       2159175b9efe66dc301f149d3d5abf8c  1371816915  28.812398  -80.883061   
4       57ff021bd3f328f8738bb535c302a31b  1371816917  44.959148  -85.884734   
...                                  ...         ...        ...         ...   
555714  9b1f753c79894c9f4b71f04581835ada  1388534347  39.946837  -91.333331   
555715  2090647dac2c89a1

In [ ]:
#dropping NULL VALUES columnwise
a.dropna(axis='columns', how='all') #how='all';which will only drop rows/columns that are all NULL VALUES

Unnamed: 0 trans_date_trans_time            cc_num  \
0                0   2020-06-21 12:14:25  2291163933867244   
1                1   2020-06-21 12:14:33  3573030041201292   
2                2   2020-06-21 12:14:53  3598215285024754   
3                3   2020-06-21 12:15:15  3591919803438423   
4                4   2020-06-21 12:15:17  3526826139003047   
...            ...                   ...               ...   
555714      555714   2020-12-31 23:59:07    30560609640617   
555715      555715   2020-12-31 23:59:09  3556613125071656   
555716      555716   2020-12-31 23:59:15  6011724471098086   
555717      555717   2020-12-31 23:59:24     4079773899158   
555718      555718   2020-12-31 23:59:34  4170689372027579   

                                    merchant        category     amt    first  \
0                      fraud_Kirlin and Sons   personal_care    2.86     Jeff   
1                       fraud_Sporer-Keebler   personal_care   29.84   Joanne   
2       fraud_Swaniawski, Nitzsche and Welch  health_fitness   41.28   Ashley   
3                          fraud_Haley Group        misc_pos   60.05    Brian   
4                      fraud_Johnston-Casper          travel    3.19   Nathan   
...                                      ...             ...     ...      ...   
555714                 fraud_Reilly and Sons  health_fitness   43.77  Michael   
555715                  fraud_Hoppe-Parisian       kids_pets  111.84     Jose   
555716                       fraud_Rau-Robel       kids_pets   86.88      Ann   
555717                 fraud_Breitenberg LLC          travel    7.99     Eric   
555718                     fraud_Dare-Marvin   entertainment   38.13   Samuel   

            last gender                       street  ...      lat      long  \
0        Elliott      M            351 Darlene Green  ...  33.9659  -80.9355   
1       Williams      F             3638 Marsh Union  ...  40.3207 -110.4360   
2          Lopez      F         9333 Valentine Point  ...  40.6729  -73.5365   
3       Williams      M  32941 Krystal Mill Apt. 552  ...  28.5697  -80.8191   
4         Massey      M     5783 Evan Roads Apt. 465  ...  44.2529  -85.0170   
...          ...    ...                          ...  ...      ...       ...   
555714     Olson      M          558 Michael Estates  ...  40.4931  -91.8912   
555715   Vasquez      M          572 Davis Mountains  ...  29.0393  -95.4401   
555716    Lawson      F   144 Evans Islands Apt. 683  ...  46.1966 -118.9017   
555717   Preston      M   7020 Doyle Stream Apt. 951  ...  44.6255 -116.4493   
555718      Frey      M     830 Myers Plaza Apt. 384  ...  35.6665  -97.4798   

        city_pop                     job         dob  \
0         333497     Mechanical engineer  1968-03-19   
1            302  Sales professional, IT  1990-01-17   
2          34496       Librarian, public  1970-10-21   
3          54767            Set designer  1987-07-25   
4           1126      Furniture designer  1955-07-06   
...          ...                     ...         ...   
555714       519            Town planner  1966-02-13   
555715     28739          Futures trader  1999-12-27   
555716      3684                Musician  1981-11-29   
555717       129            Cartographer  1965-12-15   
555718    116001             Media buyer  1993-05-10   

                               trans_num   unix_time  merch_lat  merch_long  \
0       2da90c7d74bd46a0caf3777415b3ebd3  1371816865  33.986391  -81.200714   
1       324cc204407e99f51b0d6ca0055005e7  1371816873  39.450498 -109.960431   
2       c81755dbbbea9d5c77f094348a7579be  1371816893  40.495810  -74.196111   
3       2159175b9efe66dc301f149d3d5abf8c  1371816915  28.812398  -80.883061   
4       57ff021bd3f328f8738bb535c302a31b  1371816917  44.959148  -85.884734   
...                                  ...         ...        ...         ...   
555714  9b1f753c79894c9f4b71f04581835ada  1388534347  39.946837  -91.333331   
555715  2090647dac2c89a1

In [ ]:
a.dropna(axis='rows', thresh=0)
#minimum number of non-null values for the row/column to be kept

Unnamed: 0 trans_date_trans_time            cc_num  \
0                0   2020-06-21 12:14:25  2291163933867244   
1                1   2020-06-21 12:14:33  3573030041201292   
2                2   2020-06-21 12:14:53  3598215285024754   
3                3   2020-06-21 12:15:15  3591919803438423   
4                4   2020-06-21 12:15:17  3526826139003047   
...            ...                   ...               ...   
555714      555714   2020-12-31 23:59:07    30560609640617   
555715      555715   2020-12-31 23:59:09  3556613125071656   
555716      555716   2020-12-31 23:59:15  6011724471098086   
555717      555717   2020-12-31 23:59:24     4079773899158   
555718      555718   2020-12-31 23:59:34  4170689372027579   

                                    merchant        category     amt    first  \
0                      fraud_Kirlin and Sons   personal_care    2.86     Jeff   
1                       fraud_Sporer-Keebler   personal_care   29.84   Joanne   
2       fraud_Swaniawski, Nitzsche and Welch  health_fitness   41.28   Ashley   
3                          fraud_Haley Group        misc_pos   60.05    Brian   
4                      fraud_Johnston-Casper          travel    3.19   Nathan   
...                                      ...             ...     ...      ...   
555714                 fraud_Reilly and Sons  health_fitness   43.77  Michael   
555715                  fraud_Hoppe-Parisian       kids_pets  111.84     Jose   
555716                       fraud_Rau-Robel       kids_pets   86.88      Ann   
555717                 fraud_Breitenberg LLC          travel    7.99     Eric   
555718                     fraud_Dare-Marvin   entertainment   38.13   Samuel   

            last gender                       street  ...      lat      long  \
0        Elliott      M            351 Darlene Green  ...  33.9659  -80.9355   
1       Williams      F             3638 Marsh Union  ...  40.3207 -110.4360   
2          Lopez      F         9333 Valentine Point  ...  40.6729  -73.5365   
3       Williams      M  32941 Krystal Mill Apt. 552  ...  28.5697  -80.8191   
4         Massey      M     5783 Evan Roads Apt. 465  ...  44.2529  -85.0170   
...          ...    ...                          ...  ...      ...       ...   
555714     Olson      M          558 Michael Estates  ...  40.4931  -91.8912   
555715   Vasquez      M          572 Davis Mountains  ...  29.0393  -95.4401   
555716    Lawson      F   144 Evans Islands Apt. 683  ...  46.1966 -118.9017   
555717   Preston      M   7020 Doyle Stream Apt. 951  ...  44.6255 -116.4493   
555718      Frey      M     830 Myers Plaza Apt. 384  ...  35.6665  -97.4798   

        city_pop                     job         dob  \
0         333497     Mechanical engineer  1968-03-19   
1            302  Sales professional, IT  1990-01-17   
2          34496       Librarian, public  1970-10-21   
3          54767            Set designer  1987-07-25   
4           1126      Furniture designer  1955-07-06   
...          ...                     ...         ...   
555714       519            Town planner  1966-02-13   
555715     28739          Futures trader  1999-12-27   
555716      3684                Musician  1981-11-29   
555717       129            Cartographer  1965-12-15   
555718    116001             Media buyer  1993-05-10   

                               trans_num   unix_time  merch_lat  merch_long  \
0       2da90c7d74bd46a0caf3777415b3ebd3  1371816865  33.986391  -81.200714   
1       324cc204407e99f51b0d6ca0055005e7  1371816873  39.450498 -109.960431   
2       c81755dbbbea9d5c77f094348a7579be  1371816893  40.495810  -74.196111   
3       2159175b9efe66dc301f149d3d5abf8c  1371816915  28.812398  -80.883061   
4       57ff021bd3f328f8738bb535c302a31b  1371816917  44.959148  -85.884734   
...                                  ...         ...        ...         ...   
555714  9b1f753c79894c9f4b71f04581835ada  1388534347  39.946837  -91.333331   
555715  2090647dac2c89a1

In [ ]:
#filling the NULL VALUES
a.fillna(0)

Unnamed: 0 trans_date_trans_time            cc_num  \
0                0   2020-06-21 12:14:25  2291163933867244   
1                1   2020-06-21 12:14:33  3573030041201292   
2                2   2020-06-21 12:14:53  3598215285024754   
3                3   2020-06-21 12:15:15  3591919803438423   
4                4   2020-06-21 12:15:17  3526826139003047   
...            ...                   ...               ...   
555714      555714   2020-12-31 23:59:07    30560609640617   
555715      555715   2020-12-31 23:59:09  3556613125071656   
555716      555716   2020-12-31 23:59:15  6011724471098086   
555717      555717   2020-12-31 23:59:24     4079773899158   
555718      555718   2020-12-31 23:59:34  4170689372027579   

                                    merchant        category     amt    first  \
0                      fraud_Kirlin and Sons   personal_care    2.86     Jeff   
1                       fraud_Sporer-Keebler   personal_care   29.84   Joanne   
2       fraud_Swaniawski, Nitzsche and Welch  health_fitness   41.28   Ashley   
3                          fraud_Haley Group        misc_pos   60.05    Brian   
4                      fraud_Johnston-Casper          travel    3.19   Nathan   
...                                      ...             ...     ...      ...   
555714                 fraud_Reilly and Sons  health_fitness   43.77  Michael   
555715                  fraud_Hoppe-Parisian       kids_pets  111.84     Jose   
555716                       fraud_Rau-Robel       kids_pets   86.88      Ann   
555717                 fraud_Breitenberg LLC          travel    7.99     Eric   
555718                     fraud_Dare-Marvin   entertainment   38.13   Samuel   

            last gender                       street  ...      lat      long  \
0        Elliott      M            351 Darlene Green  ...  33.9659  -80.9355   
1       Williams      F             3638 Marsh Union  ...  40.3207 -110.4360   
2          Lopez      F         9333 Valentine Point  ...  40.6729  -73.5365   
3       Williams      M  32941 Krystal Mill Apt. 552  ...  28.5697  -80.8191   
4         Massey      M     5783 Evan Roads Apt. 465  ...  44.2529  -85.0170   
...          ...    ...                          ...  ...      ...       ...   
555714     Olson      M          558 Michael Estates  ...  40.4931  -91.8912   
555715   Vasquez      M          572 Davis Mountains  ...  29.0393  -95.4401   
555716    Lawson      F   144 Evans Islands Apt. 683  ...  46.1966 -118.9017   
555717   Preston      M   7020 Doyle Stream Apt. 951  ...  44.6255 -116.4493   
555718      Frey      M     830 Myers Plaza Apt. 384  ...  35.6665  -97.4798   

        city_pop                     job         dob  \
0         333497     Mechanical engineer  1968-03-19   
1            302  Sales professional, IT  1990-01-17   
2          34496       Librarian, public  1970-10-21   
3          54767            Set designer  1987-07-25   
4           1126      Furniture designer  1955-07-06   
...          ...                     ...         ...   
555714       519            Town planner  1966-02-13   
555715     28739          Futures trader  1999-12-27   
555716      3684                Musician  1981-11-29   
555717       129            Cartographer  1965-12-15   
555718    116001             Media buyer  1993-05-10   

                               trans_num   unix_time  merch_lat  merch_long  \
0       2da90c7d74bd46a0caf3777415b3ebd3  1371816865  33.986391  -81.200714   
1       324cc204407e99f51b0d6ca0055005e7  1371816873  39.450498 -109.960431   
2       c81755dbbbea9d5c77f094348a7579be  1371816893  40.495810  -74.196111   
3       2159175b9efe66dc301f149d3d5abf8c  1371816915  28.812398  -80.883061   
4       57ff021bd3f328f8738bb535c302a31b  1371816917  44.959148  -85.884734   
...                                  ...         ...        ...         ...   
555714  9b1f753c79894c9f4b71f04581835ada  1388534347  39.946837  -91.333331   
555715  2090647dac2c89a1

In [ ]:
#again checking for the NULL VALUES in each column
a.isnull()

Unnamed: 0  trans_date_trans_time  cc_num  merchant  category    amt  \
0            False                  False   False     False     False  False   
1            False                  False   False     False     False  False   
2            False                  False   False     False     False  False   
3            False                  False   False     False     False  False   
4            False                  False   False     False     False  False   
...            ...                    ...     ...       ...       ...    ...   
555714       False                  False   False     False     False  False   
555715       False                  False   False     False     False  False   
555716       False                  False   False     False     False  False   
555717       False                  False   False     False     False  False   
555718       False                  False   False     False     False  False   

        first   last  gender  street  ...    lat   long  city_pop    job  \
0       False  False   False   False  ...  False  False     False  False   
1       False  False   False   False  ...  False  False     False  False   
2       False  False   False   False  ...  False  False     False  False   
3       False  False   False   False  ...  False  False     False  False   
4       False  False   False   False  ...  False  False     False  False   
...       ...    ...     ...     ...  ...    ...    ...       ...    ...   
555714  False  False   False   False  ...  False  False     False  False   
555715  False  False   False   False  ...  False  False     False  False   
555716  False  False   False   False  ...  False  False     False  False   
555717  False  False   False   False  ...  False  False     False  False   
555718  False  False   False   False  ...  False  False     False  False   

          dob  trans_num  unix_time  merch_lat  merch_long  is_fraud  
0       False      False      False      False       False     False  
1       False      False      False      False       False     False  
2       False      False      False      False       False     False  
3       False      False      False      False       False     False  
4       False      False      False      False       False     False  
...       ...        ...        ...        ...         ...       ...  
555714  False      False      False      False       False     False  
555715  False      False      False      False       False     False  
555716  False      False      False      False       False     False  
555717  False      False      False      False       False     False  
555718  False      False      False      False       False     False  

[555719 rows x 23 columns]

In [ ]:
#column and row count
x=a.count(axis='columns')
y=a.count()

In [ ]:
print (x)

0         23
1         23
2         23
3         23
4         23
          ..
555714    23
555715    23
555716    23
555717    23
555718    23
Length: 555719, dtype: int64


In [ ]:
print (y)

Unnamed: 0               555719
trans_date_trans_time    555719
cc_num                   555719
merchant                 555719
category                 555719
amt                      555719
first                    555719
last                     555719
gender                   555719
street                   555719
city                     555719
state                    555719
zip                      555719
lat                      555719
long                     555719
city_pop                 555719
job                      555719
dob                      555719
trans_num                555719
unix_time                555719
merch_lat                555719
merch_long               555719
is_fraud                 555719
dtype: int64


In [ ]:
#distribution of legit transactions and fraudulent transaction
a['is_fraud'].value_counts()

0    553574
1      2145
Name: is_fraud, dtype: int64

This dataset is highly unbalanced
If we enter any data points it will predict it as normal transaction

0 --> Legit Transaction
1 --> Fraudulent Transaction

In [ ]:
#separating the data for analysis
legit= a[a.is_fraud ==0]
fraud= a[a.is_fraud ==1]

In [ ]:
print(legit.shape)
print(fraud.shape)

(553574, 23)
(2145, 23)


In [ ]:
#statistical measures of the dataset for legit transaction data
legit.amt.describe()

count    553574.000000
mean         67.614408
std         152.471931
min           1.000000
25%           9.600000
50%          47.150000
75%          82.620000
max       22768.110000
Name: amt, dtype: float64

In [ ]:
#statistical measures of the dataset for fraud transaction data
fraud.amt.describe()

count    2145.000000
mean      528.356494
std       392.747594
min         1.780000
25%       214.510000
50%       371.940000
75%       907.770000
max      1320.920000
Name: amt, dtype: float64

In [ ]:
#comparing values of the both legit and fraud transactions
#for all the colums
a.groupby('is_fraud').mean()

<ipython-input-178-d5425b05a21e>:3: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  a.groupby('is_fraud').mean()


Unnamed: 0        cc_num         amt           zip        lat  \
is_fraud                                                                     
0         277997.725990  4.179643e+17   67.614408  48846.424285  38.541406   
1         242057.088112  3.854274e+17  528.356494  47862.900233  39.019971   

               long      city_pop     unix_time  merch_lat  merch_long  
is_fraud                                                                
0        -90.230495  88313.692365  1.380683e+09  38.540954  -90.230474  
1        -90.445556  64529.321678  1.379587e+09  39.018618  -90.465222

Dealing with unbalanced data

Building a sample dataset containing similar distribution of legit and fraudulent transactions
No. of legit transactions--> 553574
No. of fraudulent transactions--> 2145

joining it to make normal distribution
sample dataset

In [ ]:
#Random Sampling
legit_sample_data= legit.sample(n=2145)

In [ ]:
#Concatenating the dataframes for Normal Distribution
#Adding ROW WISE AXIS=0
#COLUMN WISE AXIS=1
concat_dataframes= pd.concat([legit_sample_data], axis=0)
c_d= concat_dataframes

In [ ]:
c_d.head()

Unnamed: 0 trans_date_trans_time            cc_num  \
384614      384614   2020-11-18 02:06:36      567868110212   
9180          9180   2020-06-23 19:55:00  3564182536169293   
251440      251440   2020-09-21 08:40:21     4640395708703   
541515      541515   2020-12-28 18:50:37  4629451965224809   
147527      147527   2020-08-11 13:37:32     4755696071492   

                                     merchant      category     amt    first  \
384614            fraud_Terry, Johns and Bins      misc_pos  122.38     Dawn   
9180                     fraud_Schimmel-Olson     kids_pets    7.66   Brenda   
251440                     fraud_Bednar Group      misc_net   34.34  Melissa   
541515    fraud_Baumbach, Strosin and Nicolas  shopping_pos    5.09    Karen   
147527  fraud_Gottlieb, Considine and Schultz  shopping_net   90.05   Robert   

           last gender                         street  ...      lat     long  \
384614   Tucker      F             789 Herring Manors  ...  29.7736 -95.4034   
9180    Johnson      F            56160 Nicholas Isle  ...  39.9934 -81.8024   
251440   Krause      F           67632 Martin Gateway  ...  30.6576 -91.6264   
541515   Warren      F  4900 Curtis Gardens Suite 952  ...  30.9416 -83.3574   
147527    Evans      M  01892 Patricia Vista Apt. 828  ...  32.9643 -96.6012   

        city_pop                                 job         dob  \
384614   2906700             Copywriter, advertising  1984-12-16   
9180        1443        Research scientist (medical)  1962-03-04   
251440      1178         Community education officer  1951-06-13   
541515     10295  Teacher, special educational needs  1997-12-26   
147527     20328                       Site engineer  1985-06-20   

                               trans_num   unix_time  merch_lat merch_long  \
384614  adc9feeb6847c621f7ba1e0a93a10d28  1384740396  29.396688 -94.785878   
9180    f99f2661d8355acb5e949ccfbbd1c4c2  1372017300  39.251661 -81.154640   
251440  5367ee840cc0a5a1e7d608626b8cddf6  1379752821  30.258607 -91.347060   
541515  50aba5d2f7d3b0a428eb134aa07c0bac  1388256637  31.516945 -82.679735   
147527  239bb55d739748c010b4918dfb807553  1376228252  32.815808 -96.979121   

        is_fraud  
384614         0  
9180           0  
251440         0  
541515         0  
147527         0  

[5 rows x 23 columns]

In [ ]:
c_d.tail()

Unnamed: 0 trans_date_trans_time            cc_num  \
433726      433726   2020-12-05 11:29:02  3598014571045296   
483056      483056   2020-12-15 14:49:02  3560725013359375   
79041        79041   2020-07-19 04:18:49  4449530933957323   
379516      379516   2020-11-16 12:32:40   213156747557083   
492226      492226   2020-12-17 23:31:21  4302480582202074   

                                    merchant        category     amt    first  \
433726                        fraud_Kuhn LLC        misc_net    3.81     Lisa   
483056          fraud_Torp, Muller and Borer  health_fitness   50.48   Brooke   
79041   fraud_Schaefer, Maggio and Daugherty   gas_transport   51.34  Felicia   
379516       fraud_Lowe, Dietrich and Erdman       kids_pets  106.86     Adam   
492226          fraud_Runte, Green and Emard  health_fitness  118.53    David   

             last gender                     street  ...      lat      long  \
433726      Clark      F          07943 Davis River  ...  35.5580  -77.2566   
483056      Smith      F  63542 Luna Brook Apt. 012  ...  31.8599 -102.7413   
79041       Mckee      F         84079 Thomas Burgs  ...  38.2427 -122.9145   
379516     Santos      M      725 Jo Trace Apt. 102  ...  34.1556 -118.2322   
492226  Rodriguez      M           821 Solis Points  ...  43.2326  -86.2492   

        city_pop                                                job  \
433726      5903                     Manufacturing systems engineer   
483056        23                                     Cytogeneticist   
79041        337                          Occupational psychologist   
379516    172817                        Advertising account planner   
492226    128715  Historic buildings inspector/conservation officer   

               dob                         trans_num   unix_time  merch_lat  \
433726  1993-10-07  a572c85667749f4768761df13a123a52  1386242942  36.245673   
483056  1969-09-15  506dc39d4affe7423ddaad31eb8393f6  1387118942  31.012722   
79041   1954-07-05  8b6291073763c7c43d7a4a28dab50fcc  1374207529  37.838564   
379516  1982-07-30  2140a09cac6c31d0a7d5e8260a653455  1384605160  34.614888   
492226  1995-05-25  6f94860d3944e2598589df36805aad9b  1387323081  43.236819   

        merch_long  is_fraud  
433726  -77.111749         0  
483056 -101.793601         0  
79041  -122.939995         0  
379516 -118.783566         0  
492226  -86.616522         0  

[5 rows x 23 columns]

In [ ]:
c_d['is_fraud'].value_counts()

0    2145
Name: is_fraud, dtype: int64

In [ ]:
c_d.groupby('is_fraud').mean()

<ipython-input-184-1561987f9891>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  c_d.groupby('is_fraud').mean()


Unnamed: 0        cc_num        amt           zip        lat  \
is_fraud                                                                    
0         275262.603263  3.983529e+17  67.044881  49124.058275  38.592409   

               long      city_pop     unix_time  merch_lat  merch_long  
is_fraud                                                                
0        -90.373258  81015.121678  1.380592e+09  38.584274  -90.375074

In [ ]:
c_d.iloc[:, 3]

384614              fraud_Terry, Johns and Bins
9180                       fraud_Schimmel-Olson
251440                       fraud_Bednar Group
541515      fraud_Baumbach, Strosin and Nicolas
147527    fraud_Gottlieb, Considine and Schultz
                          ...                  
433726                           fraud_Kuhn LLC
483056             fraud_Torp, Muller and Borer
79041      fraud_Schaefer, Maggio and Daugherty
379516          fraud_Lowe, Dietrich and Erdman
492226             fraud_Runte, Green and Emard
Name: merchant, Length: 2145, dtype: object

In [ ]:
#Spliting the into features(independent variables) and targets(dependent variable)
X= c_d.drop(columns= 'is_fraud', axis=1)
Y= c_d['is_fraud']

In [ ]:
print(X)

        Unnamed: 0 trans_date_trans_time            cc_num  \
384614      384614   2020-11-18 02:06:36      567868110212   
9180          9180   2020-06-23 19:55:00  3564182536169293   
251440      251440   2020-09-21 08:40:21     4640395708703   
541515      541515   2020-12-28 18:50:37  4629451965224809   
147527      147527   2020-08-11 13:37:32     4755696071492   
...            ...                   ...               ...   
433726      433726   2020-12-05 11:29:02  3598014571045296   
483056      483056   2020-12-15 14:49:02  3560725013359375   
79041        79041   2020-07-19 04:18:49  4449530933957323   
379516      379516   2020-11-16 12:32:40   213156747557083   
492226      492226   2020-12-17 23:31:21  4302480582202074   

                                     merchant        category     amt  \
384614            fraud_Terry, Johns and Bins        misc_pos  122.38   
9180                     fraud_Schimmel-Olson       kids_pets    7.66   
251440                     fraud_Bed

In [ ]:
print(Y)

384614    0
9180      0
251440    0
541515    0
147527    0
         ..
433726    0
483056    0
79041     0
379516    0
492226    0
Name: is_fraud, Length: 2145, dtype: int64


In [ ]:
#Spliting the data into Training and Testing data
X_train, X_test, Y_train, Y_test = tts(X, Y, test_size=0.2,stratify= Y, random_state=1)

In [ ]:
print(X.shape, X_train.shape, X_test.shape)

(2145, 22) (1716, 22) (429, 22)


In [ ]:
print(Y.shape, Y_train.shape, Y_test.shape)

(2145,) (1716,) (429,)


In [ ]:
Y_train = Y_train.astype(float)

In [ ]:
X_train['trans_date_trans_time'] = pd.to_datetime(X_train['trans_date_trans_time'])

# Extract relevant features from the timestamp column
X_train['year'] = X_train['trans_date_trans_time'].dt.year
X_train['month'] = X_train['trans_date_trans_time'].dt.month
X_train['day'] = X_train['trans_date_trans_time'].dt.day
X_train['hour'] = X_train['trans_date_trans_time'].dt.hour
X_train['minute'] = X_train['trans_date_trans_time'].dt.minute

# Drop the original timestamp column if not needed for modeling
X_train.drop('trans_date_trans_time', axis=1, inplace=True)

In [ ]:
#Training the model usning Logistic Regression for binary classification
model_lr= LogisticRegression()


In [ ]:
# Fit the model using the preprocessed features (X_train) and Y_train
model_lr.fit(X_train.values, Y_train.values)


ValueError: ignored